### PR TITLE
[ULP-1729] Add send phone message compiler

### PR DESCRIPTION
**Pull request by TSLarson**
## ✏️ Changes

Adds new send-phone-message compiler

## 🔗 References

https://auth0team.atlassian.net/browse/ULP-1729

## 🎯 Testing

✅ This has been tested in Docker using webtask-monorepo (see screenshots below)

✅ This change has unit test coverage

🚫 This change has integration test coverage

🚫 This change has been tested for performance

**Manual Testing Validation Step 1: Create Webtask with "send-phone-message" extension name**

image

**Manual Testing Validation Step 2: Execute your webtask by submitting the correct body AND adding the `auth0-extension-secret` in the header**
image

image

## 🚀 Deployment

✅ This change can support multiple releases of the code serving traffic at the same time

✅ This can be deployed any time using our continuous deploy process because it has no pre-requisites.

## 🎡 Rollout

✅ The change will be available to the complete environment at once

🚫 The change is behind a feature flag and will go through a percentage based rollout process

🚫 The change is behind a flag and customers will opt-in to the change

In order to verify that the deployment was successful we will ...

## 🔥 Rollback

This code is adding a feature that no one is currently consuming. However, if there is an issue we would rollback to the previous commit.

## 🖥 Appliance

✅  This change is compatible with the appliance.

### CODE REVIEW

File: lib/compilers/send-phone-message.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        if (typeof webtaskContext.body.recipient !== 'string') {
+            return cb(new Error('Body.recipient received by extensibility point is not a string'));
+        }
+        if (typeof webtaskContext.body.text !== 'string') {
+            return cb(new Error('Body.text received by extensibility point is not a string'));
+        }
+        if (typeof webtaskContext.body.context !== 'object') {
+            return cb(new Error('Body.context received by extensibility point is not an object'));
+        }
+        if (webtaskContext.body.context.factor_type !== ('first' || 'second')) {
```
Comment (ggoodman) [inline lib/compilers/send-phone-message.js]: This type of check won't work in JavaScript. What will end up happening is that `('first' || 'second')` will be evaluated to `'first'` and then the check will be run.

Can you please add a test for `factor_type === 'second'` and tweak the logic for this validation?

File: lib/compilers/send-phone-message.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            return cb(new Error('Body.context.factor_type received by extensibility point is not `first` or `second`'));
+        }
+        if (webtaskContext.body.context.message_type !== ('sms' || 'voice')) {
```
Comment (ggoodman) [inline lib/compilers/send-phone-message.js]: Same as before.

File: lib/compilers/send-phone-message.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            return cb(new Error('Body.context.message_type received by extensibility point is not `sms` or `voice`'));
+        }
+        if (webtaskContext.body.context.action !== ('enrollment' || 'authentication')) {
```
Comment (ggoodman) [inline lib/compilers/send-phone-message.js]: Same as before.

File: .gitignore [hunk 4]
```diff
@@ -35,3 +35,5 @@ jspm_packages
 
 # Optional REPL history
 .node_repl_history
+
+.idea
```
[the remark this answers is not in the document]
Comment (TSLarson) [reply .gitignore]: thanks for the head's up!!

File: lib/compilers/send-phone-message.js (hunk 1 above)
Comment (TSLarson) [reply lib/compilers/send-phone-message.js]: 😱 You're right! Fixing this!

Comment (TSLarson) [reply lib/compilers/send-phone-message.js]: (Still working on the happy path test case... 😬 )

**Pull request merged**